agent/proxycfg/snapshot.go

```diff
@@ -691,14 +695,85 @@ type configSnapshotAPIGateway struct {
 // Consul API Gateway into Consul core.
 //
 // FUTURE: Remove when API gateways have custom snapshot generation
-func (c *configSnapshotAPIGateway) ToIngress() configSnapshotIngressGateway {
+func (c *configSnapshotAPIGateway) ToIngress(datacenter string) (configSnapshotIngressGateway, error) {
+	// Convert API Gateway Listeners to Ingress Listeners.
+	ingressListeners := make(map[IngressListenerKey]structs.IngressListener, len(c.Listeners))
+	synthesizedChains := map[UpstreamID]*structs.CompiledDiscoveryChain{}
+	for name, listener := range c.Listeners {
+		boundListener, ok := c.BoundListeners[name]
+		if !ok {
+			// Skip any listeners that don't have a bound listener. Once the bound listener is created, this will be run again.
```

Comment:
@andrewstucki, does this track for you? I was thinking about the case where an API Gateway Listener exists, the reconciler is run, but the bound listener doesn't exist yet. I'm not sure if this case is even possible. But I figure when the bound listener is created, the reconciler will run again and this will be updated.

Reply:
So, this could be the case if something happened like we had a stale `BoundAPIGateway` -- i.e. a Gateway gets reconciled, has stuff bound to it, then renames one of its listeners. You're right that it should get re-reconciled fairly quickly, but we could potentially have a small window in which the bound gateway listeners don't match the gateway listeners themselves.

Ideally we'd also do some cursory checks like skipping routes if their protocols don't match the listener as well, just in case someone did something like changed the protocol and the resources haven't had time to reconcile yet.

Once we have a second pass at all of this with the discoverychain to GAMMA style work and potentially a bit more stable structs for our `BoundAPIGateway` data, I'd love to close these sorts of drift issues by just materializing everything onto the `BoundAPIGateway` itself so we don't have to glom multiple config entries together on the xDS side.

Reply:
Just to clarify, the reason this shouldn't ever get called when there are no bound listeners is that the `ToIngress` code only ever gets called after we have the filled-in snapshot, which I believe @nathancoleman gated on the presence of both `BoundAPIGateway` and `APIGateway` config entries.

Reply:
Ok, good. I think that makes sense. Should we wait on those cursory checks until after feature freeze?

Reply:
We can handle it in a follow-up PR, but can you `TODO` it?

Reply:
Certainly!

Reply:
> Ideally we'd also do some cursory checks like skipping routes if their protocols don't match the listener as well, just in case someone did something like changed the protocol and the resources haven't had time to reconcile yet.

I think we might not have to worry about this case because the creation of the chain with the routes will only grab routes that match the protocol.

Is this right, or am I missing something?

```go
	for _, routeRef := range boundListener.Routes {
		switch routeRef.Kind {
		case structs.HTTPRoute:
			route, ok := c.HTTPRoutes.Get(routeRef)
			if !ok {
				continue
			}
			synthesizer.AddHTTPRoute(*route)
			for _, service := range route.GetTargets() {
				id := NewUpstreamIDFromServiceName(structs.NewServiceName(service.Name, &service.EnterpriseMeta))
				if chain := c.DiscoveryChain[id]; chain != nil {
					chains = append(chains, chain)
				}
			}
		case structs.TCPRoute:
			route, ok := c.TCPRoutes.Get(routeRef)
			if !ok {
				continue
			}
			synthesizer.AddTCPRoute(*route)
			for _, service := range route.GetTargets() {
				id := NewUpstreamIDFromServiceName(structs.NewServiceName(service.Name, &service.EnterpriseMeta))
				if chain := c.DiscoveryChain[id]; chain != nil {
					chains = append(chains, chain)
				}
			}
		default:
			continue
		}
	}
```

Reply:
So, it's not so much a matter of the routes drifting, but of the fact that the listener protocol may have changed and the given routes are no longer validly bound to the gateway. Basically if we're now an HTTP listener and we generate a bunch of TCP-based discovery chains it'd be no good -- I'm not entirely sure how the ingress gateway xDS generation code handles all of this under the hood and whether it'd error out or cause some odd issues, but my thought is that we should probably throw some `listener.Protocol == route.Protocol` checks as we iterate over the listener, doing a `continue` if they don't match.